File: nim/grade-school/grade_school.nim
import std/[algorithm, sequtils]

type
  Student = tuple
    name: string
    grade: int

  School* = object
    students*: seq[Student]

proc sortedRoster(students: seq[Student]): seq[Student] =
  students.sortedByIt((it.grade, it.name))

proc roster*(s: School): seq[string] =
  sortedRoster(s.students).mapIt(it.name)

proc grade*(s: School, grade: int): seq[string] =
  sorted(s.students.filterIt(it.grade == grade).mapIt(it.name))

File: go/grade-school/grade_school.go
package school

import (
	"sort"
)

type School struct {
	grades   map[int][]string
	students map[string]bool
}

type Grade struct {
	level    int
	students []string
}

func New() *School {
	return &School{
		grades:   make(map[int][]string),
		students: make(map[string]bool),
	}
}

func (s *School) Add(student string, grade int) {
	if _, exists := s.students[student]; exists {
		return
	}
	s.students[student] = true
	s.grades[grade] = append(s.grades[grade], student)
}

func (s *School) Grade(level int) []string {
	if students, exists := s.grades[level]; exists {
		return students
	}
	return nil
}

func (s *School) Enrollment() []Grade {
	grades := make([]Grade, 0, len(s.grades))
	for level, students := range s.grades {
		sort.Slice(students, func(i, j int) bool {
			return students[i] < students[j]
		})
		grades = append(grades, Grade{level, students})
	}
	sort.Slice(grades, func(i, j int) bool {
		return grades[i].level < grades[j].level
	})
	return grades
}
